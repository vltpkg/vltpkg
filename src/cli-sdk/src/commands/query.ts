import type {
  EdgeLike,
  HumanReadableOutputGraph,
  JSONOutputGraph,
  MermaidOutputGraph,
  Node,
} from '@vltpkg/graph'
import {
  actual,
  humanReadableOutput,
  jsonOutput,
  mermaidOutput,
} from '@vltpkg/graph'
import { error } from '@vltpkg/error-cause'
import { Query } from '@vltpkg/query'
import { SecurityArchive } from '@vltpkg/security-archive'
import type { DepID } from '@vltpkg/dep-id'
import { commandUsage } from '../config/usage.ts'
import type { CommandFn, CommandUsage } from '../index.ts'
import { startGUI } from '../start-gui.ts'
import type { Views } from '../view.ts'
import type { LoadedConfig } from '../config/index.js'

export const usage: CommandUsage = () =>
  commandUsage({
    command: 'query',
    usage: [
      '',
      '<query> --view=<human | json | mermaid | gui>',
      '<query> --expect-results=<comparison string>',
    ],
    description: `List installed dependencies matching the provided query.

       The vlt Dependency Selector Syntax is a CSS-like query language that
       allows you to filter installed dependencies using a variety of metadata
       in the form of CSS-like attributes, pseudo selectors & combinators.`,
    examples: {
      [`'#foo'`]: {
        description: 'Query dependencies declared as "foo"',
      },
      [`'*:workspace > *:peer'`]: {
        description: 'Query all peer dependencies of workspaces',
      },
      [`':project > *:attr(scripts, [build])'`]: {
        description:
          'Query all direct project dependencies with a "build" script',
      },
      [`'[name^="@vltpkg"]'`]: {
        description:
          'Query packages with names starting with "@vltpkg"',
      },
      [`'*:license(copyleft) --expect-results=0'`]: {
        description: 'Errors if a copyleft licensed package is found',
      },
    },
    options: {
      'expect-results': {
        value: '[number | string]',
        description:
          'Sets an expected number of resulting items. Errors if the number of resulting items does not match the set value. Accepts a specific numeric value or a string value starting with either ">", "<", ">=" or "<=" followed by a numeric value to be compared.',
      },
      view: {
        value: '[human | json | mermaid | gui]',
        description:
          'Output format. Defaults to human-readable or json if no tty.',
      },
    },
  })

type QueryResult = JSONOutputGraph &
  MermaidOutputGraph &
  HumanReadableOutputGraph & { queryString: string }

const validateExpectedResult = (
  conf: LoadedConfig,
  edges: EdgeLike[],
): boolean => {
  const expectResults = conf.values['expect-results']
  if (expectResults?.startsWith('>=')) {
    return edges.length >= parseInt(expectResults.slice(2).trim(), 10)
  } else if (expectResults?.startsWith('<=')) {
    return edges.length <= parseInt(expectResults.slice(2).trim(), 10)
  } else if (expectResults?.startsWith('>')) {
    return edges.length > parseInt(expectResults.slice(1).trim(), 10)
  } else if (expectResults?.startsWith('<')) {
    return edges.length < parseInt(expectResults.slice(1).trim(), 10)
  } else if (expectResults) {
    return edges.length === parseInt(expectResults.trim(), 10)
  }
  return true
}

export const views = {
  json: jsonOutput,
  mermaid: mermaidOutput,
  human: humanReadableOutput,
  gui: async ({ queryString }, _, conf) => {
    await startGUI(
      conf,
      '/explore?query=' + encodeURIComponent(queryString),
    )
  },
} as const satisfies Views<QueryResult>

export const command: CommandFn<QueryResult> = async conf => {
  const monorepo = conf.options.monorepo
  const mainManifest = conf.options.packageJson.read(
    conf.options.projectRoot,
  )
  const graph = actual.load({
    ...conf.options,
    mainManifest,
    monorepo,
    loadManifests: true,
  })

  const defaultQueryString = '*'
  const queryString = conf.positionals[0]
  const securityArchive =
    queryString && Query.hasSecuritySelectors(queryString) ?
      await SecurityArchive.start({
        graph,
        specOptions: conf.options,
      })
    : undefined
  const query = new Query({
    graph,
    specOptions: conf.options,
    securityArchive,
  })

  const importers = new Set<Node>()
  const scopeIDs: DepID[] = []

  if (monorepo) {
    for (const workspace of monorepo.filter(conf.values)) {
      const w: Node | undefined = graph.nodes.get(workspace.id)
      if (w) {
        importers.add(w)
        scopeIDs.push(workspace.id)
      }
    }
  }
  if (importers.size === 0) {
    for (const importer of graph.importers) {
      importers.add(importer)
    }
  }

  const { edges, nodes } = await query.search(
    queryString || defaultQueryString,
    {
      signal: new AbortController().signal,
      scopeIDs: scopeIDs.length > 0 ? scopeIDs : undefined,
    },
  )

  if (!validateExpectedResult(conf, edges)) {
    throw error('Unexpected number of items', {
      found: edges.length,
      wanted: conf.values['expect-results'],
    })
  }

  return {
    importers,
    edges,
    nodes,
    highlightSelection: !!queryString,
    queryString: queryString || defaultQueryString,
  }
}
