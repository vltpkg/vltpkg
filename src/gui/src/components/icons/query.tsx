import { createLucideIcon } from 'lucide-react'

export const Query = createLucideIcon('Query', [
  [
    'path',
    {
      key: '1',
      fill: 'currentColor',
      fillRule: 'evenodd',
      clipRule: 'evenodd',
      strokeWidth: '0',
      d: 'M17.4858 3.771C19.0007 3.771 20.229 4.99933 20.229 6.51416C20.2289 7.31892 19.8818 8.04125 19.3293 8.54187C18.7499 9.06698 18.1714 9.71301 18.1714 10.495V13.5039C18.1714 14.286 18.7499 14.9321 19.3293 15.4573C19.8819 15.9582 20.229 16.6809 20.229 17.4858C20.229 19.0007 19.0007 20.229 17.4858 20.229C15.971 20.229 14.7427 19.0007 14.7427 17.4858C14.7427 17.4071 14.746 17.3292 14.7526 17.2522C14.8143 16.533 14.7677 15.7374 14.2572 15.227C13.7469 14.7166 12.9518 14.6707 12.2326 14.7327C12.1558 14.7393 12.0781 14.7427 11.9995 14.7427C11.921 14.7427 11.8432 14.7393 11.7664 14.7327C11.0473 14.6707 10.2521 14.7166 9.74173 15.227C9.23131 15.7374 9.1848 16.5329 9.24638 17.2522C9.25298 17.3292 9.25635 17.4071 9.25635 17.4858C9.25634 19.0006 8.0289 20.2289 6.51416 20.229C4.99933 20.229 3.771 19.0007 3.771 17.4858C3.771 16.6809 4.11806 15.9582 4.67055 15.4573C5.24998 14.932 5.82861 14.2859 5.82861 13.5038V10.4952C5.82861 9.7131 5.25001 9.06703 4.6705 8.54184C4.11812 8.04123 3.77108 7.31891 3.771 6.51416C3.771 4.99932 4.99932 3.771 6.51416 3.771C8.02886 3.77116 9.25635 4.99942 9.25635 6.51416C9.25634 6.59286 9.25296 6.67074 9.24635 6.74769C9.18464 7.46645 9.23104 8.26132 9.74115 8.77142C10.2513 9.28159 11.0466 9.32869 11.7655 9.26729C11.8426 9.26071 11.9207 9.25733 11.9995 9.25732C12.0784 9.25732 12.1565 9.26069 12.2336 9.26729C12.9525 9.32871 13.7477 9.28163 14.2578 8.77146C14.768 8.26131 14.8144 7.46643 14.7527 6.74762C14.7461 6.67069 14.7427 6.59283 14.7427 6.51416C14.7427 4.99932 15.971 3.771 17.4858 3.771ZM15.227 9.74173C14.7166 10.2521 14.6707 11.0473 14.7327 11.7664C14.7393 11.8432 14.7427 11.921 14.7427 11.9995C14.7427 12.0781 14.7393 12.1558 14.7327 12.2326C14.6707 12.9518 14.7166 13.7469 15.227 14.2572L15.8232 14.8535C15.9827 15.013 16.2264 15.0408 16.4348 14.9545C16.6432 14.8682 16.8003 14.6749 16.8003 14.4493V9.54971C16.8003 9.32414 16.6433 9.13085 16.4348 9.04455C16.2264 8.95822 15.9827 8.98604 15.8232 9.1456L15.227 9.74173ZM8.17514 9.14487C8.01601 8.98573 7.77299 8.95794 7.56503 9.04395C7.35719 9.12991 7.20068 9.32266 7.20068 9.54757V14.4502C7.20068 14.6753 7.35729 14.8682 7.56509 14.9546C7.77296 15.041 8.01604 15.0133 8.17522 14.8541L8.77146 14.2578C9.28163 13.7477 9.32871 12.9525 9.26729 12.2336C9.26069 12.1565 9.25732 12.0784 9.25732 11.9995C9.25733 11.9207 9.26071 11.8426 9.26729 11.7655C9.32869 11.0466 9.28159 10.2513 8.77142 9.74115L8.17514 9.14487Z',
    },
  ],
])
