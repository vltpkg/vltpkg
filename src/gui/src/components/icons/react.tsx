export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
