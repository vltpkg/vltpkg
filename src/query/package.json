{
  "name": "@vltpkg/query",
  "description": "Query syntax parser that retrieves items from a graph",
  "version": "0.0.0-14",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vltpkg/vltpkg.git",
    "directory": "src/query"
  },
  "tshy": {
    "selfLink": false,
    "liveDev": true,
    "dialects": [
      "esm"
    ],
    "exports": {
      "./package.json": "./package.json",
      ".": "./src/index.ts"
    }
  },
  "dependencies": {
    "@vltpkg/dep-id": "workspace:*",
    "@vltpkg/dss-parser": "workspace:*",
    "@vltpkg/error-cause": "workspace:*",
    "@vltpkg/graph": "workspace:*",
    "@vltpkg/security-archive": "workspace:*",
    "@vltpkg/semver": "workspace:*",
    "@vltpkg/types": "workspace:*",
    "p-retry": "catalog:",
    "postcss-selector-parser": "^7.1.0"
  },
  "devDependencies": {
    "@eslint/js": "catalog:",
    "@types/node": "catalog:",
    "@vltpkg/spec": "workspace:*",
    "eslint": "catalog:",
    "prettier": "catalog:",
    "tap": "catalog:",
    "tshy": "catalog:",
    "typedoc": "catalog:",
    "typescript": "catalog:",
    "typescript-eslint": "catalog:"
  },
  "license": "BSD-2-Clause-Patent",
  "engines": {
    "node": ">=22"
  },
  "scripts": {
    "format": "prettier --write . --log-level warn --ignore-path ../../.prettierignore --cache",
    "format:check": "prettier --check . --ignore-path ../../.prettierignore --cache",
    "lint": "eslint . --fix",
    "lint:check": "eslint .",
    "prepack": "tshy",
    "snap": "tap",
    "test": "tap",
    "posttest": "tsc --noEmit",
    "tshy": "tshy",
    "typecheck": "tsc --noEmit"
  },
  "tap": {
    "extends": "../../tap-config.yaml"
  },
  "prettier": "../../.prettierrc.js",
  "module": "./src/index.ts",
  "type": "module",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "default": "./src/index.ts"
      }
    }
  },
  "files": [
    "dist"
  ]
}
