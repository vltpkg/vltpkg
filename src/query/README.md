![query](https://github.com/user-attachments/assets/5b4802b7-7567-4f50-8f77-7ee398f58d43)

# @vltpkg/query

The **vlt** query syntax engine.

**[Usage](#usage)** · **[Examples](#examples)** ·
**[Supported Syntax Reference](#supported-syntax-reference)**

## Usage

```js
import { Query } from '@vltpkg/query'

const query = new Query({ nodes })
query.search(':root > *')
```

## Examples

Querying nodes from a local `node_modules` folder.

```js
import { actual } from '@vltpkg/graph'
import { Query } from '@vltpkg/query'

const graph = await actual.load({ projectRoot: process.cwd() })
const query = new Query([...graph.nodes.values()])
query.search(':root > *')
```

## Supported Syntax Reference

The vlt query syntax enable usage of css-selector-like strings to
filter packages.

Many of the common elements of the CSS language are available,
notably:

- `*` Universal selector, matches all selected items.
- `&` Nesting selector, allows for nesting selectors.
- `{}` Curly braces, when querying can be used to nest selectors.

Split by group of selectors below is the complete reference to
supported elements.

### Attribute selectors

Attribute selectors are used to match a value found in the
`package.json` metadata for each of the nodes being queried to a
arbitrary value you choose.

- `[attr]` Matches elements with an `attr` property in its
  `package.json`.
- `[attr=value]` Matches elements with a property `attr` whose value
  is exactly `value`.
- `[attr^=value]` Matches elements with a property `attr` whose value
  starts with `value`.
- `[attr$=value]` Matches elements with a property `attr` whose value
  ends with `value`.
- `[attr~=value]` Matches elements with a property `attr` whose value
  is a whitespace-separated list of words, one of which is exactly
  `value`.
- `[attr|=value]` Matches elements with a property `attr` whose value
  is either `value` or starts with `value-`.
- `[attr*=value]` Matches elements with a property `attr`.
- `[attr=value i]` Case-insensitive flag, setting it will cause any
  comparison to be case-insensitive.
- `[attr=value s]` Case-sensitive flag, setting it will cause
  comparisons to be case-sensitive, this is the default behavior.

### Class selectors

- `.prod` Matches prod dependencies to your current project.
- `.dev` Matches packages that are only used as dev dependencies in
  your current project.
- `.optional` Matches packages that are optional to your current
  project.
- `.peer` Matches peer dependencies to your current project.
- `.workspace` Matches the current project worksacpes (listed in your
  `vlt-workspaces.json` file).

### Combinators

- `>` Child combinator, matches packages that are direct dependencies
  of the previously selected nodes.
- ` ` Descendant combinator, matches all packages that are direct &
  transitive dependencies of the previously selected nodes.
- `~` Sibling combinator, matches packages that are direct
  dependencies of all dependents of the previously selected nodes.

### ID Selectors

Identifiers are a shortcut to retrieving packages by name,
unfortunately this shortcut only works for unscoped packages, with
that in mind it's advised to rely on using **Attribute selectors**
(showed above) instead.

e.g: `#foo` is the same as `[name=foo]`

### Pseudo class selectors

- `:attr(key, [attr=value])` The attribute pseudo-class allows for
  selecting packages based on nested properties of its `package.json`
  metadata. As an example, here is a query that filters only packages
  that declares an optional peer dependency named `foo`:
  `:attr(peerDependenciesMeta, foo, [optional=true])`
- `:empty` Matches packages that have no dependencies installed.
- `:has(<selector-list>)` Matches only packages that have valid
  results for the selector expression used. As an example, here is a
  query that matches all packages that have a peer dependency on
  `react`: `:has(.peer[name=react])`
- `:is(<forgiving-selector-list>)` Useful for writing large selectors
  in a more compact form, the `:is()` pseudo-class takes a selector
  list as its arguments and selects any element that can be selected
  by one of the selectors in that list. As an example, let's say I
  want to select packages named `a` & `b` that are direct dependencies
  of my project root: `:root > [name=a], :root > [name=b]` using the
  `:is()` pseudo-class, that same expression can be shortened to:
  `:root > :is([name=a], [name=b])`. Similar to the css pseudo-class
  of the same name, this selector has a forgiving behavior regarding
  its nested selector list ignoring any usage of non-existing ids,
  classes, combinators, operators and pseudo-selectors.
- `:not(<selector-list>)` Negation pseudo-class, select packages that
  do not match a list of selectors.
- `:outdated(<type>)` Matches packages that are outdated, the type
  parameter is optional and can be one of the following:
  - `any` (default) a version exists that is greater than the current
    one
  - `in-range` a version exists that is greater than the current one,
    and satisfies at least one if its parent's dependencies
  - `out-of-range` a version exists that is greater than the current
    one, does not satisfy at least one of its parent's dependencies
  - `major` a version exists that is a semver major greater than the
    current one
  - `minor` a version exists that is a semver minor greater than the
    current one
  - `patch` a version exists that is a semver patch greater than the
    current one
- `:private` Matches packages that have the property `private` set on
  their `package.json` file.
- `:semver(<value>, <function>, <custom-attribute-selector>)` Matches
  packages based on a semver value, e.g, to retrieve all packages that
  have a `version` satisfied by the semver value `^1.0.0`:
  `:semver(^1.0.0)`. It's also possible to define the type of semver
  comparison function to use by defining a second parameter, e.g:
  `:semver(^1.0.0, eq)` for an exact match, valid comparison types
  are: `eq`, `neq`, `gt`, `gte`, `lt`, `lte`, `satisfies` (default). A
  third parameter allows for specifying a different `package.json`
  property to use for the comparison, e.g:
  `:semver(^22, satisfies, :attr(engines, [node]))` for comparing the
  value of `engines.node`.
- `:type(registry|file|git|remote|workspace)` Matches packages based
  on their type, e.g, to retrieve all git dependencies: `:type(git)`.

### Pseudo Elements

- `:project` Returns both the root node (as defined below) along with
  any workspace declared in your project.
- `:root` Returns the root node, that represents the package defined
  at the top-level `package.json` of your project folder.
- `:scope` Returns the current scope of a given selector
