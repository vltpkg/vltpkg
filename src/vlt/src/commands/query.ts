import { fileURLToPath } from 'node:url'
import {
  actual,
  EdgeLike,
  humanReadableOutput,
  jsonOutput,
  mermaidOutput,
  Node,
  NodeLike,
} from '@vltpkg/graph'
import { Query } from '@vltpkg/query'
import { commandUsage } from '../config/usage.js'
import { type CliCommandUsage, CliCommandFn } from '../types.js'
import { startGUI } from '../start-gui.js'

export const usage: CliCommandUsage = () =>
  commandUsage({
    command: 'query',
    usage: ['', '<query> --view=[human | json | mermaid | gui]'],
    description:
      'List installed dependencies matching the provided query.',
    examples: {
      [`'#foo'`]: {
        description: 'Query packages with the name "foo"',
      },
      [`'*.workspace > *.peer'`]: {
        description: 'Query all peer dependencies of workspaces',
      },
      [`':project > *:attr(scripts, [build])'`]: {
        description:
          'Query all direct project dependencies with a "build" script',
      },
      [`'[name^="@vltpkg"]'`]: {
        description:
          'Query packages with names starting with "@vltpkg"',
      },
    },
    options: {
      view: {
        value: '[human | json | mermaid | gui]',
        description:
          'Output format. Defaults to human-readable or json if no tty.',
      },
    },
  })

export const view = {
  json: jsonOutput,
  mermaid: mermaidOutput,
  human: humanReadableOutput,
}

export type CommandResult = {
  importers: Set<Node>
  edges: EdgeLike[]
  nodes: NodeLike[]
  highlightSelection: boolean
}

export const command: CliCommandFn = async (
  conf,
  assetsDir: string = fileURLToPath(
    import.meta.resolve('@vltpkg/gui'),
  ),
) => {
  const monorepo = conf.options.monorepo
  const mainManifest = conf.options.packageJson.read(
    conf.options.projectRoot,
  )
  const graph = actual.load({
    ...conf.options,
    mainManifest,
    monorepo,
    loadManifests: true,
  })

  const defaultQueryString = '*'
  const queryString = conf.positionals[0]
  const query = new Query({ graph })
  const { edges, nodes } = await query.search(
    queryString || defaultQueryString,
  )

  const importers = new Set<Node>()
  if (monorepo) {
    for (const workspace of monorepo.filter(conf.values)) {
      const w: Node | undefined = graph.nodes.get(workspace.id)
      if (w) importers.add(w)
    }
  }
  if (importers.size === 0) {
    for (const importer of graph.importers) {
      importers.add(importer)
    }
  }

  if (conf.values.view === 'gui') {
    await startGUI({
      assetsDir,
      conf,
      startingRoute:
        '/explore?query=' +
        encodeURIComponent(queryString || defaultQueryString),
    })
    return
  }

  return {
    /* c8 ignore next */
    defaultView: !process.stdout.isTTY ? 'json' : 'human',
    result: {
      importers,
      edges,
      nodes,
      highlightSelection: !!queryString,
    },
  }
}
