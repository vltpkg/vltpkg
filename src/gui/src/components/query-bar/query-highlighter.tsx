import { forwardRef } from 'react'
import { cn } from '@/lib/utils.ts'
import { QueryToken } from '@/components/query-bar/query-token.tsx'
import type { ParsedSelectorToken } from '@vltpkg/query'

interface QueryHighlighterProps {
  query: string
  parsedTokens: ParsedSelectorToken[]
  className?: string
}

export const QueryHighlighter = forwardRef<
  HTMLDivElement,
  QueryHighlighterProps
>(({ query, parsedTokens, className }, ref) => {
  if (!query) return null

  let q = query
  let curr = 0
  const result = []
  while (q) {
    const astItem = parsedTokens[curr]
    const token = astItem?.token
    const type = astItem?.type

    if (!token) {
      const segment = {
        type: 'text' as ParsedSelectorToken['type'],
        token: q,
      }
      result.push(segment)
      break
    }

    if (q.startsWith(token)) {
      const segment = {
        type,
        token,
      }

      result.push(segment)

      q = q.slice(token.length)
      curr++
    } else {
      const [unrecognizedToken] = q.split(token)
      if (!unrecognizedToken) break
      const segment = {
        type: 'string' as ParsedSelectorToken['type'],
        token: unrecognizedToken,
      }
      result.push(segment)
      q = q.slice(unrecognizedToken.length)
      continue
    }
  }

  return (
    <div
      ref={ref}
      className={cn(
        'scrollbar-hidden pointer-events-none absolute inset-0 left-[2.556rem] top-[0.01rem] flex items-center overflow-x-auto pr-32',
        className,
      )}>
      <div className="flex whitespace-nowrap">
        {result.map((segment, idx) => (
          <QueryToken variant={segment.type} key={idx}>
            {segment.token}
          </QueryToken>
        ))}
      </div>
    </div>
  )
})
