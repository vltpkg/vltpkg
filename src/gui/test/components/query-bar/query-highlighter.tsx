import { vi, test, expect, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { QueryHighlighter } from '@/components/query-bar/query-highlighter.jsx'
import { Query } from '@vltpkg/query'

vi.mock('@/components/query-bar/query-token.jsx', () => ({
  QueryToken: 'gui-query-token',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

test('query-highlighter render default', () => {
  const mockQuery = ':root > :not(#react, [name="react-sass"])'
  const mockParsedTokens = Query.getQueryTokens(mockQuery)

  const Container = () => {
    return (
      <QueryHighlighter
        query={mockQuery}
        parsedTokens={mockParsedTokens}
      />
    )
  }

  const { container } = render(<Container />)
  expect(container.innerHTML).toMatchSnapshot()
})
