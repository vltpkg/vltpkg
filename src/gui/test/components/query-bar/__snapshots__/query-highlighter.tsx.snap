// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`query-highlighter render default 1`] = `

<div class="scrollbar-hidden pointer-events-none absolute inset-0 left-[2.556rem] top-[0.01rem] flex items-center overflow-x-auto pr-32">
  <div class="flex whitespace-nowrap">
    <gui-query-token variant="pseudo">
      :root
    </gui-query-token>
    <gui-query-token variant="combinator">
      &gt;
    </gui-query-token>
    <gui-query-token variant="pseudo">
      :not(
    </gui-query-token>
    <gui-query-token variant="id">
      #react
    </gui-query-token>
    <gui-query-token variant="selector">
      ,
    </gui-query-token>
    <gui-query-token variant="attribute">
      [name="react-sass"]
    </gui-query-token>
    <gui-query-token variant="pseudo">
      )
    </gui-query-token>
  </div>
</div>

`;
